gexpect_test.go

```diff
@@ -236,3 +236,88 @@ func TestRegexTimeoutWithOutput(t *testing.T) {
 		t.Fatalf("Didn't find %v in output: %v", searchPattern, out)
 	}
 }
+
+func TestRegexFindNoExcessBytes(t *testing.T) {
+	t.Logf("Testing Regular Expressions returning output with no excess strings")
+	repeats := 100
+	tests := []struct {
+		desc           string
+		loopBody       string
+		searchPattern  string
+		expectFullTmpl string
+		unmatchedData  string
+	}{
+		{
+			desc:           `matching lines line by line with $ at the end of the regexp`,
+			loopBody:       `echo "prefix: ${i} line"`,
+			searchPattern:  `(?m)^prefix:\s+(\d+) line\s??$`,
+			expectFullTmpl: `prefix: %d line`,
+			unmatchedData:  "\n",
+			// the "$" char at the end of regexp does not
+			// match the \n, so it is left as an unmatched
+			// data
+		},
+		{
+			desc:           `matching lines line by line with \n at the end of the regexp`,
+			loopBody:       `echo "prefix: ${i} line"`,
+			searchPattern:  `(?m)^prefix:\s+(\d+) line\s??\n`,
+			expectFullTmpl: `prefix: %d line`,
+			unmatchedData:  "",
+		},
+		{
+			desc:           `matching chunks in single line chunk by chunk`,
+			loopBody:       `echo -n "a ${i} b"`,
+			searchPattern:  `a\s+(\d+)\s+b`,
+			expectFullTmpl: `a %d b`,
+			unmatchedData:  "",
+		},
+	}
+	seqCmd := fmt.Sprintf("`seq 1 %d`", repeats)
+	shCmdTmpl := fmt.Sprintf(`sh -c 'for i in %s; do %%s; done'`, seqCmd)
+	for _, tt := range tests {
+		t.Logf("Test: %s", tt.desc)
+		shCmd := fmt.Sprintf(shCmdTmpl, tt.loopBody)
+		t.Logf("Running command: %s", shCmd)
+		p, err := Spawn(shCmd)
+		if err != nil {
+			t.Fatalf("Cannot exec shell script: %v", err)
+		}
+		defer func() {
+			if err := p.Wait(); err != nil {
+				t.Fatalf("shell script didn't terminate correctly: %v", err)
+			}
+		}()
+		for i := 1; i <= repeats; i++ {
+			matches, output, err := p.ExpectRegexFindWithOutput(tt.searchPattern)
+			if err != nil {
+				t.Fatalf("Failed to get the match number %d: %v", i, err)
+			}
+			if len(matches) != 2 {
+				t.Fatalf("Expected only 2 matches, got %d", len(matches))
```

Comment:
In this case we expect the `matches` slice to have the whole match (like "prefix: 1 line") and a submatch for the number only ("1").